### PR TITLE
Fluid Typography: ensure global styles preset fluid sizes are calculated correctly

### PR DESCRIPTION
**Pull request by ramonjd**
Resolves https://github.com/WordPress/gutenberg/issues/44752

## What

This PR is from https://github.com/WordPress/gutenberg/pull/44761 to fix https://github.com/WordPress/gutenberg/issues/44752, and fixes another bug that forces preset values selected in the site editor are calculated to use any available preset fluid parameters.

It also ensure that the control UI for the font size picker works as intended:

https://user-images.githubusercontent.com/6458278/194701949-e3cc7e30-c265-4bda-9fe4-c8010c3f20cb.mp4

https://user-images.githubusercontent.com/6458278/194701957-6c1977c2-03cd-4135-b7e5-7e23c340ac69.mp4

## How

By pre-calculating fluid values before building the fontsizepicker controls so that the clamp() value is saved.

Custom font size values will be calculated by [getStyleDeclarations](https://github.com/WordPress/gutenberg/pull/44761/files#diff-1654d2da9097b71ac77806bbc505483b794371e8a2ff3ba152ec8610c2663ec6R281)

## Testing Instructions

Try with TT3 or the example JSON:

<details>

<summary>Example theme.json</summary>

```json
{
	"version": 2,
	"settings": {
		"appearanceTools": true,
		"typography": {
			"fluid": true,
			"fontSizes": [
				{
					"fluid": {
						"min": "0.875rem",
						"max": "1rem"
					},
					"size": "1rem",
					"slug": "small"
				},
				{
					"fluid": {
						"min": "1rem",
						"max": "1.125rem"
					},
					"size": "1.125rem",
					"slug": "medium"
				},
				{
					"size": "1.9rem",
					"slug": "not-so-large",
					"fluid": {
						"min": "1.9rem",
						"max": "2rem"
					}
				},
				{
					"size": "3.5rem",
					"slug": "large",
					"fluid": {
						"min": "3rem",
						"max": "5rem"
					}
				},
				{
					"size": "2.25rem",
					"slug": "x-large",
				},
				{
					"size": "10rem",
					"slug": "xx-large",
					"fluid": {
						"min": "4rem",
						"max": "20rem"
					}
				},
				{
					"size": "14rem",
					"slug": "Colossal",
					"fluid": {
						"min": "8rem",
						"max": "30rem"
					}
				}
			]
		}
	}
}


```

</details>

1. Go to the site editor
2. For global styles (elements, blocks etc) apply a preset font size and save.
3. Ensure that a fluid font size is applied in the editor and in the frontend.
4. Repeat but apply a custom font size.

### CODE REVIEW

File: packages/components/src/font-size-picker/utils.ts [hunk 1]
```diff
@@ -142,13 +145,26 @@ export function getToggleGroupOptions(
...
-		CUSTOM_FONT_SIZE_OPTION
+
+	const fontSizeOptions = getFontSizeOptions(
+		useSelectControl,
+		fontSizes,
+		disableCustomFontSizes
 	);
+
+	const selectedOption = fontSizeOptions
+		? // @ts-ignore @TODO fix types for fontSizeOptions. Array.find() works on the same type only.
```
Comment (mikachan) [inline packages/components/src/font-size-picker/utils.ts]: It looks like this may be related to a Typescript issue with union types: https://github.com/microsoft/TypeScript/issues/44373

I don't think there's currently a fix but thought it might be useful to just drop the link here for future reference.

**Review by mikachan — APPROVED** [review]
This is working great for me!

I can now see the correct fluid font size based on the available preset parameters. For example, in TT3 I can see when I set all H1s to the XXL font size, the clamp calculation with the same values from theme.json is used:

Preset in theme.json:
image

Output:
image

I can also confirm that the XXL label is shown correctly in the Global Styles UI:

image

File: packages/components/src/font-size-picker/utils.ts (hunk 1 above)
Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts]: Thanks for finding that. 

Yeah, it was doing my head in. I tried all sorts of unions before realizing that the impenetrable 100-line-long Typescript error was really wah wahing for the same object 🙃 

I'll add a note as you say. Cheers!

**Comment by noisysocks** [comment]
You'll need a `CHANGELOG.md` update for the `@wordpress/components` changes.

**Review by andrewserong — APPROVED** [review]
Thanks for the fix here @ramonjd, I think this is a good pragmatic fix for the problem at hand — it solves the problem by ensuring that the clampified values are saved, instead of the raw non-clampified values, which ensures that when `fluid` is set to `true`, the fluid font-sizes are in use.

However, I'm also wondering if the behaviour here reveals something slightly unexpected in how preset font size values are saved in global styles versus in the post editor. In the post editor, when a preset is selected, we use the `fontSize` attribute to store the slug of the preset. But in global styles the real value of the preset is used instead of storing a reference to the CSS variable, which creates an interesting problem when we try toggling the `fluid` setting. I'll try to walk through:

With `fluid` set to `false`, set the H1 element to use `xx-large` based on the sample `theme.json` settings. Then, update the `theme.json` file to set `fluid` to `true`, and reload the site editor. Notice that global styles now thinks that a custom value is in use, because the value differs to the clampified value:

| Fluid set to `false` | Fluid set to `true` (existing value looks like a custom value now) |
| --- | --- |
| image | image |

If we go from the other direction (set a fluid typography preset and then toggle off the `fluid` value), then when we reload, the font size picker displays an empty custom value field (because the current clampified value cannot be displayed in the UnitControl). So it appears empty:



The last problem is that the clampified values don't appear to work with the value label in the select view. Note that the grey values next to the slugs aren't visible in the fluid view in global styles, however they are visible in the post editor view:

| Global styles | Post editor |
| --- | --- |
| image | image |

---

Now, all that said — I think some of these things are things we can potentially look into beyond 6.1. Because user values for global styles are saved at the per-theme level, and it's unlikely that someone will be toggling `fluid`, in practice, I think it's doubtful that folks will really run into the issue of using real values all that frequently.

So, please excuse the long rambling explanation above as I think through the trade-offs! For expedience, and to resolve the bug at hand, I think this is a good fix to go with (and it addresses the bug at hand). For 6.2, though, it might be nice to for us to dig in a little further to how global styles saves font sizes, and see if we can add some better flexibility / referencing the CSS variables, along with exposing UI controls surrounding fluid typography.

What do you think?

**Comment by ramonjd** [comment]
Thanks for testing @noisysocks and @andrewserong 

> With fluid set to false, set the H1 element to use xx-large based on the sample theme.json settings. Then, update the theme.json file to set fluid to true, and reload the site editor. Notice that global styles now thinks that a custom value is in use, because the value differs to the clampified value

Yeah it appears inconsistent. I wonder if we could categorize it, as you already mention, as an idiosyncrasy of the site editor/presets.

There is a slightly weak argument that says, well, if you're toggling fluidity on and off for individual font sizes, any presently saved value will be custom since there's no equivalent in the presets collection.

But it would still be neat to have a value in the UI. 

Perhaps we could at least have a fallback placeholder value. 

> For 6.2, though, it might be nice to for us to dig in a little further to how global styles saves font sizes, and see if we can add some better flexibility / referencing the CSS variables, along with exposing UI controls surrounding fluid typography.

I agree. And also with your observation that, for now, there won't be much toggling of individual font sizes so maybe it's moot for 6.1. 

> The last problem is that the clampified values don't appear to work with the value label in the select view. Note that the grey values next to the slugs aren't visible in the fluid view in global styles, however they are visible in the post editor view:

I'm thinking that resolving this might be doable. I'll have a play and get back to you.

**Comment by ramonjd** [comment]
> The last problem is that the clampified values don't appear to work with the value label in the select view.

It looks like this has always been true for any non-simple CSS value: https://github.com/WordPress/gutenberg/blob/9d5cfe70697e25db9c73d35cab0f7378e6c6390a/packages/components/src/font-size-picker/utils.ts#L119

We pass a `__experimentalHint` to `<CustomSelectControl />`.

So, even in the days before fluid typography we didn't know how to deal with `var()` or `clamp()` values.

Maybe in the future we should have a more meaningful hint depending on the type of value:

Screen Shot 2022-10-10 at 12 57 39 pm

Long story short: I think we can punt it.

**Comment by andrewserong** [comment]
> Long story short: I think we can punt it.

Sounds good 👍

File: packages/components/src/font-size-picker/utils.ts [hunk 2]
```diff
@@ -134,6 +136,7 @@ export function getToggleGroupOptions(
 		return {
 			key: slug,
 			value: size,
+			size,
```
Comment (noisysocks) [inline packages/components/src/font-size-picker/utils.ts:139]: Why do we need both `size` and `value`?

File: packages/components/src/font-size-picker/utils.ts [hunk 3]
```diff
@@ -142,13 +145,27 @@ export function getToggleGroupOptions(
 
 export function getSelectedOption(
 	fontSizes: FontSize[],
-	value: FontSizePickerProps[ 'value' ]
+	value: FontSizePickerProps[ 'value' ],
+	useSelectControl: boolean,
+	disableCustomFontSizes: boolean = false
 ): FontSizeOption {
```
Comment (noisysocks) [inline packages/components/src/font-size-picker/utils.ts:151]: The return type here is `FontSizeOption` but your changes make it so that the return type is `FontSizeSelectOption | FontSizeToggleGroupOption`. This should be a type error but I guess it's not because of the `// @ts-ignore`.

Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:151]: Oh good point. That was an oversight. The `// @ts-ignore` is for something else entirely. I'll update.

Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:151]: On the other PR coz I just merged 😄

File: packages/components/src/font-size-picker/utils.ts (hunk 2 above)
Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:139]: Long story :)

It's the way `ToggleGroupControl` and `ToggleGroupControlOption` work.

The latter uses a `value` prop for the font size, the former `size`. See: https://github.com/WordPress/gutenberg/blob/d66796f043b484e3103ec8d9ce41ed839a62038f/packages/components/src/toggle-group-control/types.ts#L126

We need both so the current size is displayed correctly in the options and the `ToggleGroupControl` header.

File: packages/components/src/font-size-picker/utils.ts (hunk 3 above)
Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:151]: Here's what I need to do to make TS shut up

```patch
diff --git a/packages/components/src/font-size-picker/types.ts b/packages/components/src/font-size-picker/types.ts
index ec193f51bf..1df66bb922 100644
--- a/packages/components/src/font-size-picker/types.ts
+++ b/packages/components/src/font-size-picker/types.ts
@@ -92,6 +92,7 @@ export type FontSizeSelectOption = Pick< FontSizeOption, 'size' > & {
 
 export type FontSizeToggleGroupOption = {
 	key: string;
+	slug: string;
 	value: number | string;
 	label: string;
 	name: string;
diff --git a/packages/components/src/font-size-picker/utils.ts b/packages/components/src/font-size-picker/utils.ts
index 02427e0419..b59c3609a2 100644
--- a/packages/components/src/font-size-picker/utils.ts
+++ b/packages/components/src/font-size-picker/utils.ts
@@ -135,6 +135,7 @@ export function getToggleGroupOptions(
 	return optionsArray.map( ( { slug, size, name }, index ) => {
 		return {
 			key: slug,
+			slug,
 			value: size,
 			size,
 			label: labelAliases[ index ],
@@ -148,7 +149,7 @@ export function getSelectedOption(
 	value: FontSizePickerProps[ 'value' ],
 	useSelectControl: boolean,
 	disableCustomFontSizes: boolean = false
-): FontSizeOption {
+): FontSizeOption | FontSizeToggleGroupOption {
 	if ( ! value ) {
 		return DEFAULT_FONT_SIZE_OPTION;
 	}

```

File: packages/components/src/font-size-picker/utils.ts (hunk 2 above)
Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:139]: What I wrote above is complete rubbish. Ignore me.

Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:139]: > What I wrote above is complete rubbish. Ignore me.

Actually, I was right and I didn't know it. The `size` has to stay.

Comment (ramonjd) [reply packages/components/src/font-size-picker/utils.ts:139]: > Actually, I was right and I didn't know it. The size has to stay.

Actually I'm wrong and I didn't know it. It's the `headerHint` I need to change. Ignore everything.

**Comment by ndiego** [comment]
@ramonjd @noisysocks does this need to be backported to 6.1?

**Comment by ramonjd** [comment]
> @ramonjd @noisysocks does this need to be backported to 6.1?

Thanks @ndiego This PR has already been merged into https://github.com/WordPress/gutenberg/pull/44761, which needs to be backported.

Therefore this PR can be safely ignored for the purposes of backporting. 

🍺

**Comment by mikachan** [comment]
@ramonjd, I'm seeing the 'undefined' label again with the XXL font size on WP 6.1 RC1 and Gutenberg trunk. Happy to create another issue.

image

**Comment by mikachan** [comment]
I think I have a fix for the 'undefined' issue described above. It looks like it's only a problem when the theme does not provide `name` keys for each font size preset: https://github.com/WordPress/gutenberg/pull/45041
